yt_dlp/extractor/rai.py

```diff
@@ -96,12 +98,128 @@ def _extract_relinker_info(self, relinker_url, video_id):
         if not formats and geoprotection is True:
             self.raise_geo_restricted(countries=self._GEO_COUNTRIES)
 
+        formats.extend(self._create_http_urls(relinker_url, formats))
+
         return dict((k, v) for k, v in {
             'is_live': is_live,
             'duration': duration,
             'formats': formats,
         }.items() if v is not None)
 
+    def _create_http_urls(self, relinker_url, fmts):
+        _RELINKER_REG = r'https?://(?P<host>[^/]+?)/(?:i/)?(?P<extra>[^/]+?)/(?P<path>.+?)/(?P<id>\d+)(?:_(?P<quality>[\d\,]+))?(?:\.mp4|/playlist\.m3u8).+?'
+        _MP4_REG = r'https?://(?P<cdn_id>creativemedia[x]?|download)(?P<cdn_num>\d+).+?/\d+(?:_\d+)?\.mp4$'
+        _MP4_TMPL = 'https://%s%s-rai-it.akamaized.net/%s%s/%s%s.mp4'
+        _QUALITY = {
+            # tbr: w, h
+            '250': [352, 198],
+            '400': [512, 288],
+            '700': [512, 288],
+            '800': [700, 394],
+            '1200': [736, 414],
+            '1800': [1024, 576],
+            '2400': [1280, 720],
+            '3200': [1440, 810],
+            '3600': [1440, 810],
+            '5000': [1920, 1080],
+            '10000': [1920, 1080],
+        }
+
+        def get_location(url, headers={}):
+            _url = compat_urllib_parse_urlparse(url)
+            conn = compat_http_client.HTTPConnection(_url.netloc)
+            headers.update({'Connection': 'close'})
+            try:
+                conn.request(
+                    'HEAD', '%s?%s' % (_url.path, _url.query),
+                    headers=headers)
+            except Exception:
+                return None
+
+            resp = conn.getresponse()
+
+            if resp.status == 200:
+                return True
+            elif resp.status == 302:
+                return resp.getheader('Location')
+            elif resp.status == 404:
+                return None
+            return None
+
```

Comment:
Use this only if you cannot do what you want with `_download_webpage_handle` or even `YoutubeDL.urlopen`

I do see some other small stuff I can review, but this is the most important part

Reply:
Premise: The steaming server can be tricked into giving the direct http mp4 url (most of the times) by simply sending a clean http request without user-agent header.

1. So I need a method that does this trick. I tried using `headers={'User-Agent': ''}` in the usual methods but it didn't work and so I created this method that do a very clean http request.

2. I this request fails and return a m3u8 manifest, I manually create the direct mp4 url and test if it exist (you can see it at line 191) and I needed a method that didn't pollute the console while testing a bunch of made up urls, most of them would return 404 until I find one that return 200, and I don't need the user to see a lot of "warnings" on the screen

So this method does this two things.

Reply:
From my understanding, doing it this way will cause network options like `--no-check-certificate` and `--proxy` to not work correctly

> So I need a method that does this trick. I tried using headers={'User-Agent': ''} in the usual methods but it didn't work and so I created this method that do a very clean http request.

I'll try to figure out why this doesn't work

Reply:
I think it's because not sending the user-agent is not "compliant" to http rules or something and so Youtube-dl always sends "correct" http requests to be compliant.

But I need to sent a not "standard" request to the server to trick it and give me what I want,

Reply:
I tried to get `headers={'User-Agent': None'}` to work, but that needs modification to `YoutubeDL._opener` and doing so in a python2 compatible way is quite difficult. So we can proceed with your method instead.

Let me know once you have finalized the rest and I will do a full review